src/client/serialize/daos_serialize.c

```diff
@@ -8,13 +8,89 @@
 
 /* for user attr dataset */
 struct usr_attr {
-	char	*attr_name;
-	hvl_t	attr_val;
+    char    *attr_name;
+    hvl_t   attr_val;
 };
 
+/* struct obj_id type defined for hdf5 attribute. The DAOS_PROP_CO_ROOTS
+ * points to an array of daos_obj_id_t, which is a struct that contains
+ * hi, lo uint64_t's. In order to write these to an hdf5 file a similar
+ * type has to be defined to describe/store it to hdf5
+ */
+struct obj_id {
+    uint64_t hi;
+    uint64_t lo;
+};
+
+int
+serialize_roots(hid_t file_id, struct daos_prop_entry *entry, const char *prop_str)
+{
+	int				rc = 0;
+	hid_t				status = 0;
+	struct daos_prop_co_roots	*roots;
+	struct obj_id			root_oids[4];
+	hsize_t				attr_dims[1];
+	/* HDF5 returns non-negative identifiers if successfully opened/created */
+	hid_t				attr_dtype = -1;
+	hid_t				attr_dspace = -1;
+	hid_t				usr_attr = -1;
+	int				i = 0;
+
+	if (entry == NULL || entry->dpe_val_ptr == NULL) {
+        	D_GOTO(out, rc = -DER_INVAL);
+    	}
+
+	roots = entry->dpe_val_ptr;
+	attr_dims[0] = 4;
+
+	for (i = 0; i < 4; i++) {
+		root_oids[i].hi = roots->cr_oids[i].hi;
+		root_oids[i].lo = roots->cr_oids[i].lo;
+	}
+
+	attr_dtype = H5Tcreate(H5T_COMPOUND, sizeof(struct obj_id));
+	if (attr_dtype < 0) {
+		D_ERROR("failed to create attribute datatype");
+		D_GOTO(out, rc = -DER_MISC);
+	}
+	status = H5Tinsert(attr_dtype, "hi", HOFFSET(struct obj_id, hi), H5T_NATIVE_UINT64);
+	if (status < 0) {
+		D_ERROR("failed to insert oid hi");
+		D_GOTO(out, rc = -DER_MISC);
+	}
+	status = H5Tinsert(attr_dtype, "lo", HOFFSET(struct obj_id, lo), H5T_NATIVE_UINT64);
+	if (status < 0) {
+		D_ERROR("failed to insert oid low");
+		D_GOTO(out, rc = -DER_MISC);
+	}
+
+	attr_dspace = H5Screate_simple(1, attr_dims, NULL);
+	if (attr_dspace < 0) {
+        	D_ERROR("failed to create version attribute");
+		D_GOTO(out, rc = -DER_MISC);
+	}
+	usr_attr = H5Acreate2(file_id, prop_str, attr_dtype, attr_dspace, H5P_DEFAULT, H5P_DEFAULT);
+	if (usr_attr < 0) {
+		D_ERROR("failed to create attribute");
+		D_GOTO(out, rc = -DER_MISC);
+	}
+	status = H5Awrite(usr_attr, attr_dtype, root_oids);
+	if (status < 0) {
+		D_ERROR("failed to write attribute");
+		D_GOTO(out, rc = -DER_MISC);
+	}
+out:
+	if (usr_attr >= 0)
+		H5Aclose(usr_attr);
+	if (attr_dtype >= 0)
+        	H5Tclose(attr_dtype);
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs